### PR TITLE
eif: Add EIF header architecture flag

### PR DESCRIPTION
**Pull request by alcioa**
When building an EIF image, inspect the source docker
image and based on its architecture (i.e. amd64 vs aarch64)
set the flag accordingly in the EIF header flags.

This flag shall be cleared for amd64 and toggled for aarch64.

Signed-off-by: Alexandru Ciobotaru <alcioa@amazon.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: enclave_build/src/lib.rs [hunk 1]
```diff
@@ -181,11 +183,22 @@ impl<'a> Docker2Eif<'a> {
             return Err(Docker2EifError::LinuxkitExecError);
         }
 
+        let arch = self.docker.architecture().map_err(|e| {
+            eprintln!("Docker error: {:?}", e);
+            Docker2EifError::DockerError
+        })?;
+
+        let flags: u16 = if arch == "arm64" {
```
Comment (lucianperju) [inline enclave_build/src/lib.rs:191]: arm64 should be a const defined somewhere

File: enclave_build/src/docker.rs [hunk 2]
```diff
@@ -14,6 +14,10 @@ use tempfile::NamedTempFile;
 use tokio::prelude::{Future, Stream};
 use url::Url;
 
+/// Docker inspect architecture constants
+pub const DOCKER_ARCH_ARM64: &str = "arm64";
+pub const DOCKER_ARCH_AMD64: &str = "amd64";
```
Comment (petreeftime) [inline enclave_build/src/docker.rs:19]: I'd go for the more commonly used aarch64 and x86_64 instead, especially since it's difficult to see the difference between arm64 and amd64.

Comment (alcioa) [reply enclave_build/src/docker.rs:19]: That was the intention but `docker inspect` returns these constants.

File: eif_defs/src/lib.rs [hunk 3]
```diff
@@ -18,6 +18,10 @@ use std::mem::size_of;
 
 pub const EIF_MAGIC: [u8; 4] = [46, 101, 105, 102]; // .eif in ascii
 pub const MAX_NUM_SECTIONS: usize = 32;
+/// EIF Header Flags
+/// bits 1  : Architecture - toggled for aarch64, cleared for x86_64
+/// bits 2-8: Unused
+pub const EIF_HDR_ARCH_ARM64: u16 = 0x1;
```
Comment (petreeftime) [inline eif_defs/src/lib.rs:24]: What if we ever need add RISC-V or something else? We should have a separate arch field.

Comment (petreeftime) [reply eif_defs/src/lib.rs:24]: Given that we might not want to add another version of EIF right now, I think this is fine.

File: Cargo.lock [hunk 4]
```diff
@@ -391,6 +391,7 @@ version = "0.1.0"
 dependencies = [
  "base64",
  "clap",
+ "eif_defs",
```
Comment (alexggh) [inline Cargo.lock:394]: why is this needed now ?

Comment (alcioa) [reply Cargo.lock:394]: Because `enclave_build` now depends on `eif_defs` and `Cargo.lock` required update.

File: enclave_build/src/lib.rs [hunk 5]
```diff
@@ -181,11 +184,23 @@ impl<'a> Docker2Eif<'a> {
             return Err(Docker2EifError::LinuxkitExecError);
         }
 
+        let arch = self.docker.architecture().map_err(|e| {
+            eprintln!("Docker error: {:?}", e);
+            Docker2EifError::DockerError
+        })?;
+
+        let flags = match arch.as_str() {
```
Comment (alexggh) [inline enclave_build/src/lib.rs:192]: I think the right abstraction is to pass the arch string to EifBuilder and let that object take the decision.
And this way you avoid making enclave_build dependent on eif_defs.

Comment (alcioa) [reply enclave_build/src/lib.rs:192]: This would require switching `eif_utils` dependent on `enclave_build/docker.rs`. Let's pass the flag instead of the arch string from here for now.

**Review by bercarug — APPROVED** [review]

**Review by lucianperju — APPROVED** [review]
